ark/codegen.cpp

```diff
@@ -0,0 +1,437 @@
+
+   private:
+    std::string def_op(const Json &op_json, size_t task_id, size_t op_idx);
+
+    std::string def_task(const Json &task_json);
+
+    std::string def_channels(int world_size);
+
+    std::string task_seq(size_t proc_b, size_t proc_e, size_t proc_s,
+                         size_t proc_cur, size_t task_b, size_t task_e,
+                         size_t task_s, size_t task_gran, size_t num_slots,
+                         size_t slot_num_warps, size_t slot_sram_bytes,
+                         size_t task_id);
+
+    std::string resource_group(const Json &rg_json, const Json &task_infos,
+                               const Range<size_t> &proc_range);
+
+    std::string sync_process_range(const Range<size_t> &ranges, int state_id);
+
+   protected:
+    friend class CodeGenerator;
+
+    std::map<size_t, size_t> buffer_id_to_offset_;
+    std::string name_;
+    int rank_;
+    int world_size_;
+    size_t num_procs_;
+    size_t num_warps_per_proc_;
+    std::string code_;
+};
+
+CodeGenerator::Impl::Impl(const Json &plan,
+                          const std::map<size_t, size_t> &buffer_id_to_offset,
+                          const std::string &name)
+    : buffer_id_to_offset_(buffer_id_to_offset), name_(name) {
+    rank_ = plan.at("Rank");
+    world_size_ = plan.at("WorldSize");
+    num_procs_ = plan.at("NumProcessors");
+    num_warps_per_proc_ = plan.at("NumWarpsPerProcessor");
+
+    std::stringstream definitions_ss;
+    for (auto &task_json : plan.at("TaskInfos")) {
+        definitions_ss << this->def_task(task_json);
+    }
+
+    if (world_size_ > 1) {
+        definitions_ss << this->def_channels(world_size_);
+    }
+
+    std::map<Range<size_t>, SyncStateInfo> sync_state_info;
+
+    auto get_state_id = [&sync_state_info](const Range<size_t> &range) {
+        auto it = sync_state_info.find(range);
+        if (it == sync_state_info.end()) {
+            size_t id = sync_state_info.size();
+            sync_state_info.emplace(range, SyncStateInfo(id));
+            return id;
+        }
+        return it->second.id;
+    };
+
+    std::list<Range<size_t>> unsynced;
+    std::stringstream body_ss;
+    size_t pg_idx = 0;
+    for (auto &pg : plan.at("ProcessorGroups")) {
+        Range<size_t> proc_range(pg["ProcessorRange"][0],
+                                 pg["ProcessorRange"][1]);
+        size_t begin = *proc_range.begin();
+        size_t end = *proc_range.end();
+        if (end == begin) continue;
+
+        if (pg_idx > 0) {
+            bool need_sync = false;
+            auto it = unsynced.begin();
+            while (it != unsynced.end()) {
+                auto &range = *it;
+                auto intersec = proc_range.intersection(range);
+                if (intersec.empty()) {
+                    it++;
+                    continue;
+                }
+                // no need to sync if range is a part of proc_range.
+                if (intersec.size() < range.size()) {
```

Comment:
For these code, do you mean we don't need sync if proc_ranges are same for two ops? What if op2 depends on op1 with same proc_range but thread-block with same id deal with different part of data. Is this possible?

Reply:
Sorry for confusion, actually we do sync when proc_ranges are the same. I updated comments in [1fbf38b](https://github.com/microsoft/ark/pull/205/commits/1fbf38b4dd839f97199ade30a90959c5145f1a86).